tls/s2n_record_read_cbc.c

```diff
@@ -86,6 +86,7 @@ int s2n_record_parse_cbc(
 
     /* Subtract the padding length */
     gt_check(en.size, 0);
+    gte_check(payload_length, en.data[en.size - 1] + 1);
     payload_length -= (en.data[en.size - 1] + 1);
```

Comment:
minor nit - you might want to calculate the length to subtract once and use it in the following two lines - you could even make a macro to do it for you:

```
#define s2n_subtract_underflow(n, diff) do { \
  gte_check((n), (diff)); \
  (n) -= (diff); \
} while (0)
```

then we can just invoke:

```
s2n_subtract_underflow(payload_length, en.data[en.size - 1] + 1);
```

Reply:
What is the purpose of: "do{ } while(0)"?

Reply:
makes your macro look like a function, so that the semicolon on use isn't spurious.

see how our other macros use this pattern.  it also creates scope if you want to use a temporary variable.

Reply:
+1 on the macro.  It does reduce the risk of having the values be different. But, it still ends up evaluating diff twice. Might be better to do something like

```
#define s2n_subtract_underflow(n, diff) do { \
  typeof(diff) __temp_diff = (diff);
  gte_check((n), (__temp_diff)); \
  (n) -= (__temp_diff); \
} while (0)

```

Reply:
thanks for addressing that - when i first wrote this i had something similar, and now i don't know why i took it out.

Reply:
Nit on the macro name 🙂: could we call it `CHECKED_SUB_ASSIGN`, instead?

The other name doesn't really communicate that it assigns the outcome to the lhs. I also prefer macros to be all caps, which makes it clearer that there's potential flow control included with the invocation.

Reply:
naming is hard :)  i have no strong feelings on the one i hacked up, but @camshaft has a good suggestion! 👍

Reply:
Also I called out in another comment that we should try using the built-in functions for this https://github.com/awslabs/s2n/pull/2308#issuecomment-698564748

Reply:
Hmm is it better to just use this function? https://github.com/awslabs/s2n/blob/main/utils/s2n_safety.c#L228-L234?

Reply:
good find :)  perhaps we should move all those simple things to the header file and mark as `static inline`, the compiler will just optimize the callstack out for us too.... or perhaps it already does with the hiding of symbols change....

but either way, yes, lets use what exists!

Reply:
As long as you're using `uint32_t` then that function works. The nice thing about the macro is it's generic over the numeric type so it can be used in several contexts.

Reply:
we could also just replace the function with the new and improved generic macro... ;)

Reply:
Unfortunately, that would require non-standard statement exprs (see https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html)